parser/src/transforms/constant_propagation.rs

```diff
@@ -0,0 +1,486 @@
+                                *expr = ScalarExpr::Const(Span::new(span, value[idx]));
+                            }
+                            ref ty => panic!(
+                                "invalid constant reference, expected scalar access, got {:?}",
+                                ty
+                            ),
+                        },
+                        ConstantExpr::Matrix(value) => match sym.access_type {
+                            AccessType::Matrix(row, col) => {
+                                *expr = ScalarExpr::Const(Span::new(span, value[row][col]));
+                            }
+                            ref ty => panic!(
+                                "invalid constant reference, expected scalar access, got {:?}",
+                                ty
+                            ),
+                        },
+                    }
+                }
+                ControlFlow::Continue(())
+            }
+            // Fold constant expressions
+            ScalarExpr::Binary(ref mut binary_expr) => {
+                let span = binary_expr.span();
+                match self.try_fold_binary_expr(binary_expr) {
+                    Ok(maybe_folded) => {
+                        if let Some(folded) = maybe_folded {
+                            *expr = ScalarExpr::Const(Span::new(span, folded));
+                        }
+                        ControlFlow::Continue(())
+                    }
+                    Err(err) => ControlFlow::Break(err),
+                }
+            }
+            // While calls cannot be constant folded, arguments can be
+            ScalarExpr::Call(ref mut call) => self.visit_mut_call(call),
+            // This cannot be constant folded
+            ScalarExpr::BoundedSymbolAccess(_) => ControlFlow::Continue(()),
+        }
+    }
+
+    fn visit_mut_expr(&mut self, expr: &mut Expr) -> ControlFlow<InvalidConstantError> {
+        let span = expr.span();
+        match expr {
+            // Already constant
+            Expr::Const(_) => ControlFlow::Continue(()),
+            // Lift to `Expr::Const` if the scalar expression is constant
+            //
+            // We deal with symbol accesses directly, as they may evaluate to an aggregate constant
+            Expr::SymbolAccess(ref mut access) => {
+                let constant_value = match access.name {
+                    // Possibly a reference to a constant declaration
+                    ResolvableIdentifier::Resolved(ref qid) => {
+                        self.global.get(qid).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Possibly a reference to a local bound to a constant
+                    ResolvableIdentifier::Local(ref id) => {
+                        self.local.get(id).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Other identifiers cannot possibly be constant
+                    _ => None,
+                };
```

Comment:
this code is duplicated in the function above, so I'd prefer to use a helper in both cases